### ISSUE TITLE
ARSubsystemManager Removed

### ISSUE THREAD

**Issue by milamila33** [shaqian/TF-Unity-ARFoundation#4]
Hi, tried to update since ARSubsystemManager is Removed, but did not succeed (new to coding). Kindly appreciate any help.

**Comment by Spiderguy-F** [comment]
The setup is a little more complicated, you can copy it from one of the sample scenes though. Here is an updated version of the CameraImage.cs, that works fine for me with MobileNet Classifier option:
`using System;
using System.Collections;
using System.Collections.Generic;
using Unity.Collections.LowLevel.Unsafe;
using UnityEngine;
using UnityEngine.UI;
using UnityEngine.XR.ARFoundation;
using UnityEngine.XR.ARSubsystems;

/// <summary>
/// This component tests getting the latest camera image
/// and converting it to RGBA format. If successful,
/// it displays the image on the screen as a RawImage
/// and also displays information about the image.
///
/// This is useful for computer vision applications where
/// you need to access the raw pixels from camera image
/// on the CPU.
///
/// This is different from the ARCameraBackground component, which
/// efficiently displays the camera image on the screen. If you
/// just want to blit the camera texture to the screen, use
/// the ARCameraBackground, or use Graphics.Blit to create
/// a GPU-friendly RenderTexture.
///
/// In this example, we get the camera image data on the CPU,
/// convert it to an RGBA format, then display it on the screen
/// as a RawImage texture to demonstrate it is working.
/// This is done as an example; do not use this technique simply
/// to render the camera image on screen.
/// </summary>
public class CameraController : MonoBehaviour
{
    [SerializeField]
    [Tooltip("The ARCameraManager which will produce frame events.")]
    ARCameraManager m_CameraManager;

    /// <summary>
    /// Get or set the <c>ARCameraManager</c>.
    /// </summary>
    public ARCameraManager cameraManager
    {
        get { return m_CameraManager; }
        set { m_CameraManager = value; }
    }
    public ARRaycastManager m_raycastManager;
    [SerializeField]
    RawImage m_RawImage;

    /// <summary>
    /// The UI RawImage used to display the image on screen.
    /// </summary>
    public RawImage rawImage
    {
        get { return m_RawImage; }
        set { m_RawImage = value; }
    }

    [SerializeField]
    Text m_ImageInfo;

    [SerializeField]
    Text m_ObjectFoundInfo;

    [Header("Camera Image Handler")]
    [SerializeField]
    TextAsset model;

    [SerializeField]
    TextAsset labels;

    [SerializeField]
    GameObject indicator;

    Classifier classifier;
    Detector detector;

    private IList outputs;
    private GameObject apple;
    /// <summary>
    /// The UI Text used to display information about the image on screen.
    /// </summary>
    public Text imageInfo
    {
        get { return m_ImageInfo; }
        set { m_ImageInfo = value; }
    }

    void OnEnable()
    {
        if (m_CameraManager != null)
        {
            m_CameraManager.frameReceived += OnCameraFrameReceived;
        }
        InitTF();
        InitIndicator();
    }

    void OnDisable()
    {
        if (m_CameraManager != null)
        {
            m_CameraManager.frameReceived -= OnCameraFrameReceived;
        }
    }

    unsafe void OnCameraFrameReceived(ARCameraFrameEventArgs eventArgs)
    {
        // Attempt to get the latest camera image. If this method succeeds,
        // it acquires a native resource that must be disposed (see below).
        XRCameraImage image;
        if (!cameraManager.TryGetLatestImage(out image))
        {
            return;
        }

        // Display some information about the camera image
        m_ImageInfo.text = string.Format(
            "Image info:\n\twidth: {0}\n\theight: {1}\n\tplaneCount: {2}\n\ttimestamp: {3}\n\tformat: {4}",
            image.width, image.height, image.planeCount, image.timestamp, image.format);

        // Once we have a valid XRCameraImage, we can access the individual image "planes"
        // (the separate channels in the image). XRCameraImage.GetPlane provides
        // low-overhead access to this data. This could then be passed to a
        // computer vision algorithm. Here, we will convert the camera image
        // to an RGBA texture and draw it on the screen.

        // Choose an RGBA format.
        // See XRCameraImage.FormatSupported for a complete list of supported formats.
        var format = TextureFormat.RGBA32;

        if (m_Texture == null || m_Texture.width != image.width || m_Texture.height != image.height)
        {
            m_Texture = new Texture2D(image.width, image.height, format, false);
        }

        // Convert the image to format, flipping the image across the Y axis.
        // We can also get a sub rectangle, but we'll get the full image here.
        var conversionParams = new XRCameraImageConversionParams(image, format, CameraImageTransformation.MirrorY);

        // Texture2D allows us write directly to the raw texture data
        // This allows us to do the conversion in-place without making any copies.
        var rawTextureData = m_Texture.GetRawTextureData<byte>();
        try
        {
            image.Convert(conversionParams, new IntPtr(rawTextureData.GetUnsafePtr()), rawTextureData.Length);
        }
        finally
        {
            // We must dispose of the XRCameraImage after we're finished
            // with it to avoid leaking native resources.
            image.Dispose();
        }

        // Apply the updated texture data to our texture
        m_Texture.Apply();

        // Set the RawImage's texture so we can visualize it.
        m_RawImage.texture = m_Texture;

        //RunTF(m_Texture);
    }

    public void RunTF()
    {
        if (m_Texture == null)
            print("there was no m_Texture");
        RunTF(m_Texture);
    }

    public void InitTF()
    {
        // MobileNet
        classifier = new Classifier(model, labels, output: "MobilenetV1/Predictions/Reshape_1");

        // SSD MobileNet
        //detector = GetComponent<Detector>();
        //detector.InitDetector(model, labels, input: "image_tensor");

        // Tiny YOLOv2
        //detector = new Detector(model, labels, DetectionModels.YOLO,
        //width: 416,
        //height: 416,
        //mean: 0,
        //std: 255);
    }
    public void InitIndicator()
    {
        apple = Instantiate(indicator, new Vector3(0, 0, 0), Quaternion.identity);
        apple.transform.localScale = new Vector3(0.0004f, 0.0004f, 0.0004f);
        apple.SetActive(false);
    }

    public void RunTF(Texture2D texture)
    {
        // MobileNet
        outputs = classifier.Classify(texture, angle: 90, threshold: 0.05f);
        //print("Dictionary: " + outputs + ", Count: " + outputs.Count);
        // SSD MobileNet
        //outputs = detector.Detect(m_Texture, angle: 90, threshold: 0.6f);

        // Tiny YOLOv2
        //outputs = detector.Detect(m_Texture, angle: 90, threshold: 0.1f);

        // Draw AR apple
        for (int i = 0; i < outputs.Count; i++)
        {
            //var output = outputs[i] as Dictionary<string, object>;
            print("Type: " + outputs[i].GetType());
            printDict((KeyValuePair<string, float>)outputs[i]);
            //if (output["detectedClass"].Equals("apple"))
            //{
            //    DrawApple(output["rect"] as Dictionary<string, float>);
            //    break;
            //}
        }
    }

    void printDict(KeyValuePair<string, float> pair)
    {
        print("Key: " + pair.Key + ", Value: " + pair.Value);
    }

    private void DrawApple(Dictionary<string, float> rect)
    {
        var xMin = rect["x"];
        var yMin = 1 - rect["y"];
        var xMax = rect["x"] + rect["w"];
        var yMax = 1 - rect["y"] - rect["h"];

        var pos = GetPosition((xMin + xMax) / 2 * Screen.width, (yMin + yMax) / 2 * Screen.height);

        apple.SetActive(true);
        apple.transform.position = pos;
    }

    private Vector3 GetPosition(float x, float y)
    {
        var hits = new List<ARRaycastHit>();

        m_raycastManager.Raycast(new Vector3(x, y, 0), hits, TrackableType.Planes);

        if (hits.Count > 0)
        {
            var pose = hits[0].pose;

            return pose.position;
        }

        return new Vector3();
    }

    Texture2D m_Texture;
}
`

**Comment by MarcinKulwicki** [comment]
Still don't known how to set up scene with this script

**Comment by Spiderguy-F** [comment]
Use the sample scenes from the AR Foundation git. Check out how there raycasting scenes (any ones with manual object placement) is setup and copy it, or just drag and drop all assets from the AR Foundation git into your project to get direct access to their scenes. 
There needs to be a RaycastManager present and it needs to be linked to the m_raycastManager in the editor.
If this doesn't clarify things, please give me more information than a general idk.

**Comment by lordalcol** [comment]
Hi @Spiderguy-F , I am also trying to update this project to AR Foundation 2.0. Is this something you may have already done, perhaps?

**Comment by Spiderguy-F** [comment]
All you need to do is go into the package manager and click update on all the packages with the download symbol. I figured you did this bc. your SubsystemManager was gone.

**Comment by lordalcol** [comment]
Yes, I got everything to compile, I had to copy some code from the AR Foundation examples as you suggested. Unfortunately it is not working (on Android). SSD_MOBILENET is failing the initialization with the error `Error Unity: TFException: Op type not registered 'NonMaxSuppressionV3' in binary running on localhost. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be donebefore importing the graph, as contrib ops are lazily registered when the module is first accessed.
` , while YOLO is showing labels and rects that are completely wrong. Would you know why that is happening?
